compiler/rustc_trait_selection/src/traits/select/mod.rs

```diff
@@ -1308,6 +1308,12 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
         other: &EvaluatedCandidate<'tcx>,
         needs_infer: bool,
     ) -> bool {
+        // HACK(eddyb) remove instruction-counting noise from `-Z self-profile`.
+        #[cfg(target_arch = "x86_64")]
+        unsafe {
+            asm!("mfence", options(nostack));
+        }
```

Comment:
@davidtwco @nikomatsakis Just it doesn't get lost, I don't think we should merge with this, but I also am not sure what to do here. Do you think it would be reasonable to limit this to when `-Z self-profile` is enabled?

It's also not easy to test the impact of this fence.

Reply:
I'm not too sure either. I personally wouldn't have a problem with this being limited to when `-Z self-profile` is enabled.

Reply:
Checking for `-Zself-profile` is probably more expensive than unconditionally running `mfence`.

Reply:
I'm fine leaving it in as it is for now. We can analyze its impact once perftest multithreaded rustc